### ISSUE TITLE
Constructor like syntax is parsed as method if the name is wrong

### ISSUE THREAD

**Issue by scheglov** [dart-lang/sdk#31001]
```dart
class A {
  B() : super();
}
```

```kernel
library;
import self as self;
import "dart:core" as core;

class A extends core::Object {
  default constructor •() → void
    : super core::Object::•()
    ;
  abstract method B() → dynamic;
}
static const field dynamic #errors = const <dynamic>["file:///test.dart:2:16: Error: Can't have initializers: B\n  B() : super();\n               ^"]/* from null */;
```

There is inconsistency between Analyzer parser and Fasta.

Analyzer reports an error and treats `B() : super()` as `A() : super()` because there is `:` after parameters.

Fasta says that this is a method, and it should not have constructor initializers.

I think Analyzer's behavior is a bit better because we lose less nodes when we recover this way.

**Comment by peter-ahe-google** [comment]
I agree that the analyzer's way is better. Later today, I hope the two of us can walk through how to recover from errors with constructors and factories. What Fasta is currently doing doesn't work really well. See also #30981.

Furthermore, if it had been a factory, Fasta would create a factory called "B", as if the user had written `factory A.B() {...}`. I think it would be better to assume the user meant `factory A() { ... }`.

**Comment by peter-ahe-google** [comment]
The factory problem is reported as #30999.

**Comment by bwilkerson** [comment]
I believe that this CL fixes the AstBuilder: https://dart-review.googlesource.com/c/sdk/+/25640.

**Comment by jensjoha** [comment]
This currently gives this error message:
```
t.dart:2:3: Error: The name of a constructor must match the name of the enclosing class.
  B() : super();
  ^
```

and compiles like this:
```
import self as self;
import "dart:core" as core;

class A extends core::Object {
  constructor B() → self::A
    : super core::Object::•()
    ;
}
```

(i.e. as a named constructor called B)

There's probably still an inconsistency so I'll leave it open, but I don't think the CFE recovery here is bad.
